### PR TITLE
Kulaev Evgeny. Laboratory work №1. Option №2

### PR DESCRIPTION
**Pull request by zcotti**
A plugin that adds functions without conditions to the `always_inline` attribute and outputs it

### CODE REVIEW

File: clang/test/lab1/kulaev_zhenya/test.cpp [hunk 1]
```diff
@@ -0,0 +1,18 @@
+// RUN: %clang_cc1 -load %llvmshlibdir/AttrubuteAlwaysPlugin%pluginext -plugin always_inlines-plugin %s 1>&1 | FileCheck %s
+
+// CHECK: __attribute__((always_inline)) sum
```
Comment (aobolensk) [inline clang/test/lab1/kulaev_zhenya/test.cpp]: Please, check that attribute actually appeared

File: clang/labs/lab1/kulaev_zhenya/AttrubuteAlways.cpp [hunk 2]
```diff
@@ -0,0 +1,58 @@
+#include "clang/AST/ASTConsumer.h"
+#include "clang/AST/Attr.h"
+#include "clang/AST/RecursiveASTVisitor.h"
+#include "clang/Frontend/CompilerInstance.h"
+#include "clang/Frontend/FrontendPluginRegistry.h"
+
+using namespace clang;
```
Comment (aobolensk) [inline clang/labs/lab1/kulaev_zhenya/AttrubuteAlways.cpp]: Please, use clang namespace explicitly, avoid using "using namespace *"

**Comment by aobolensk** [comment]
Please, update the branch to go through updated clang-format check

File: clang/labs/lab1/kulaev_zhenya/AttrubuteAlways.cpp [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+class AlwaysInlinePlugin : public PluginASTAction {
+public:
+  std::unique_ptr<clang::ASTConsumer>
+  CreateASTConsumer(CompilerInstance &Compiler,
+                    llvm::StringRef) override {
+    return std::make_unique<AlwaysInlineConsumer>(&Compiler.getASTContext());
+  }
+  bool ParseArgs(const CompilerInstance &Compiler,
+                 const std::vector<std::string> &args) override {
```
Comment (m-ly4) [inline clang/labs/lab1/kulaev_zhenya/AttrubuteAlways.cpp]: Could you please fix warnings printed by clang-tidy test: https://github.com/NN-complr-tech/llvm-nnsu-2024/actions/runs/8250927820/job/22566738976?pr=6

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: clang/labs/lab1/kulaev_zhenya/AttrubuteAlways.cpp (hunk 3 above)
Comment (m-ly4) [reply clang/labs/lab1/kulaev_zhenya/AttrubuteAlways.cpp]: There are still some warnings in clang-tidy that can be fixed.

**Comment by zcotti** [comment]
1) Fixed previous errors
2) Fixed clang-format
3) Fixed a bug with clang-tidy

@m-ly4  But my windows crashed, and apparently not because of me, Polozov Vladislav had a similar error, how can I fix it?

File: clang/labs/lab1/kulaev_zhenya/AttrubuteAlways.cpp (hunk 3 above)
Comment (zcotti) [reply clang/labs/lab1/kulaev_zhenya/AttrubuteAlways.cpp]: We will fix it)

File: clang/test/lab1/kulaev_zhenya/test.cpp [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+// CHECK: checkAlw not suitable for the attribute
+void checkAlw(int A, int B) {
+  {
+    if (A > B) {
+      A = B;
+    }
+  }
+}
+
+// CHECK: Added attribute always_inline in checkEmpty
```
Comment (aobolensk) [inline clang/test/lab1/kulaev_zhenya/test.cpp]: Please add check for the attribute, not for logs that you print by yourself

**Review by aobolensk — APPROVED** [review]

**Review by m-ly4 — APPROVED** [review]

**Pull request merged**